client/src/main/java/com/alibaba/nacos/client/config/impl/ClientWorker.java

```diff
@@ -129,14 +129,12 @@ public void addListeners(String dataId, String group, List<? extends Listener> l
      * @param listener listener
      */
     public void removeListener(String dataId, String group, Listener listener) {
-        group = null2defaultGroup(group);
-        CacheData cache = getCache(dataId, group);
-        if (null != cache) {
-            cache.removeListener(listener);
-            if (cache.getListeners().isEmpty()) {
-                agent.removeCache(dataId, group);
-            }
-        }
+        String finalGroup = null2defaultGroup(group);
+        Optional<CacheData> cacheDataOptional = getCache(dataId, finalGroup);
+        cacheDataOptional.ifPresent(cacheData -> {
+            cacheData.removeListener(listener);
+            agent.removeCache(dataId, finalGroup);
```

Comment:
the old logic will judge cache.listeners is empty, why remove this logic?

Reply:
In order to relieve the concurrency problem, the logic is here `com.alibaba.nacos.client.config.impl.ClientWorker#removeCacheIfEmptyListener(java.lang.String, java.lang.String)`
![image](https://user-images.githubusercontent.com/7919984/101426934-df2aad80-3938-11eb-8398-9c0b97327852.png)